crates/bevy_ecs/src/event/trigger.rs

```diff
@@ -0,0 +1,461 @@
+    event::{EntityEvent, Event},
+    observer::{CachedObservers, TriggerContext},
+    traversal::Traversal,
+    world::DeferredWorld,
+};
+use bevy_ptr::PtrMut;
+use core::marker::PhantomData;
+
+/// [`Trigger`] determines _how_ an [`Event`] is triggered when [`World::trigger`](crate::world::World::trigger) is called.
+/// This decides which [`Observer`](crate::observer::Observer)s will run, what data gets passed to them, and the order they will
+/// be executed in.
+///
+/// Implementing [`Trigger`] is "advanced-level" territory, and is generally unnecessary unless you are developing highly specialized
+/// [`Event`] trigger logic.
+///
+/// Bevy comes with a number of built-in [`Trigger`] implementations (see their documentation for more info):
+/// - [`GlobalTrigger`]: The [`Event`] derive defaults to using this
+/// - [`EntityTrigger`]: The [`EntityEvent`] derive defaults to using this
+/// - [`PropagateEntityTrigger`]: The [`EntityEvent`] derive uses this when propagation is enabled.
+/// - [`EntityComponentsTrigger`]: Used by Bevy's [component lifecycle events](crate::lifecycle).
+///
+/// # Safety
+///
+/// Implementing this properly is _advanced_ soundness territory! Implementers must abide by the following:
+///
+/// - The `E`' [`Event::Trigger`] must be constrained to the implemented [`Trigger`] type, as part of the implementation.
+///   This prevents other [`Trigger`] implementations from directly deferring to your implementation, which is a very easy
+///   soundness misstep, as most [`Trigger`] implementations will invoke observers that are developed _for their specific [`Trigger`] type_.
+///   Without this constraint, something like [`GlobalTrigger`] could be called for _any_ [`Event`] type, even one that expects a different
+///   [`Trigger`] type. This would result in an unsound cast of [`GlobalTrigger`] reference.
+///   This is not expressed as an explicit type constraint,, as the `for<'a> Event::Trigger<'a>` lifetime can mismatch explicit lifetimes in
+///   some impls.
+/// - Read, understand, and abide by the lifetime constraints defined in the section below:
+///
+/// To understand why this must be unsafe, we must think carefully about the lifetimes involved.
+///
+/// The core challenge is that the lifetime of the `&mut E::Trigger<'_>` that we want to create
+/// within the [`ObserverRunner`] (which calls the [`Observer`](crate::observer::Observer)) may not be the same as the lifetime provided
+/// by [`Event::Trigger<'a>`].
+///
+/// If the lifetimes are the same, then we can safely create a `&mut E::Trigger<'_>` from the [`PtrMut`]
+/// passed to the observer runner function, and pass that to the observer system inside of ['On'](crate::observer::On).
+///
+/// If the lifetimes are not the same, then we must be careful to ensure that the `&mut E::Trigger<'_>` we create
+/// does not outlive the [`PtrMut`] that was passed to the observer runner function.
+/// Failing to do so could lead to use-after-free bugs.
```

Comment:
I don't think this really explains the situation. After all the discussion I think there are just two situations:

- the lifetime is unused, and hence this is trivially sound
- the lifetime is used, but cannot be the same as the original one, so the different one must not be exposed to safe code.

For the second part I believe this just falls back to `On` not exposing the first of its two lifetimes.